pennylane/beta/tapes/tape.py

```diff
@@ -936,3 +936,387 @@ def execute_device(self, params, device):
+
+        return tuple(allowed_param_methods.values())
+
+    def numeric_pd(self, idx, device, params=None, **options):
+        """Evaluate the gradient of the tape with respect to
+        a single trainable tape parameter using numerical finite-differences.
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (~.Device, ~.QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+
+        Keyword Args:
+            h=1e-7 (float): finite difference method step size
+            order=1 (int): The order of the finite difference method to use. ``1`` corresponds
+                to forward finite differences, ``2`` to centered finite differences.
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+                measurement statistics
+        """
+        if params is None:
+            params = np.array(self.get_parameters())
+
+        order = options.get("order", 1)
+        h = options.get("h", 1e-7)
+
+        shift = np.zeros_like(params)
+        shift[idx] = h
+
+        if order == 1:
+            # Forward finite-difference.
+            # Check if the device has already be pre-computed with
+            # unshifted parameter values, to avoid redundant evaluations.
+            y0 = options.get("y0", None)
+
+            if y0 is None:
+                y0 = np.asarray(self.execute_device(params, device))
+
+            y = np.array(self.execute_device(params + shift, device))
+            return (y - y0) / h
+
+        if order == 2:
+            # central finite difference
+            shift_forward = np.array(self.execute_device(params + shift / 2, device))
+            shift_backward = np.array(self.execute_device(params - shift / 2, device))
+            return (shift_forward - shift_backward) / h
+
+        raise ValueError("Order must be 1 or 2.")
+
+    def device_pd(self, device, params=None, **options):
+        """Evaluate the gradient of the tape with respect to
+        a single trainable tape parameter by querying the provided device.
+
+        Args:
+            device (~.Device, ~.QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+        """
+        # pylint:disable=unused-argument
+        if params is None:
+            params = np.array(self.get_parameters())
+
+        saved_parameters = self.get_parameters()
+
+        # temporarily mutate the in-place parameters
+        self.set_parameters(params)
```

Comment:
Just to understand, what are the roles of the different parameters here? A fully built tape has fixed parameters, but when the jacobian is evaluated at some other parameters, we need to temporarily replace them? Is that a weird artefact of eager execution in a setting that is much more suitable to some kind of computational graph, where inputs are just placeholders?

Reply:
Yep, exactly. Due to needing to support torch/tf/autograd, we need to support both
```python
>>> tape.jacobian(dev) # use the existing parameters
>>> tape.jacobian(dev, params=...) # use new parameters
```

They all have their own weird intricacies! In particular, autograd forces you to pass parameters to the wrapped function, otherwise it will claim that the output has an unconnected gradient. Torch and TF, on the other hand, do not care